File: Examenes-EDD-1/2021-Septiembre/Enunciado/SparseMatrix.java
package sparseMatrix;

import java.util.TreeMap;
import java.util.List;

public class SparseMatrix {
    private final int rows;
    private final int cols;
    private final TreeMap<Index, Integer> values;

    public SparseMatrix(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.values = new TreeMap<>();
    }

    // Devuelve el valor en una posición (privado)
    private int value(Index idx) {
        return values.getOrDefault(idx, 0);
    }

    // Modifica el valor en una posición (privado)
    private void update(Index idx, int val) {
        if (val == 0) {
            values.remove(idx);
        } else {
            values.put(idx, val);
        }
    }

    // Devuelve un índice validado
    private Index index(int r, int c) {
        if (r < 0 || r >= rows || c < 0 || c >= cols)
            throw new IndexOutOfBoundsException("Índice fuera de rango");
        return new Index(r, c);
    }

    // Pública: set
    public void set(int r, int c, int val) {
        Index idx = index(r, c);
        update(idx, val);
    }

    // Pública: get
    public int get(int r, int c) {
        Index idx = index(r, c);
        return value(idx);
    }

    // Pública: suma de dos matrices
    public SparseMatrix add(SparseMatrix other) {
        // Verificar dimensiones, lanzar excepción si no coinciden
        // Implementación eficiente
        return null; // por implementar
    }

    // Pública: traspuesta
    public SparseMatrix transpose() {
        return null; // por implementar
    }

    // Pública: fromList
    public static SparseMatrix fromList(int rows, int cols, List<Integer> data) {
        if (data.size() % 3 != 0)
            throw new IllegalArgumentException("La lista no es múltiplo de 3");
        SparseMatrix mat = new SparseMatrix(rows, cols);
        for (int i = 0; i < data.size(); i += 3) {
            int r = data.get(i);
            int c = data.get(i + 1);
            int v = data.get(i + 2);
            mat.set(r, c, v);
        }
        return mat;
    }
}

File: Examenes-EDD-5/2021-Septiembre/Haskell/SparseMatrix.hs
module SparseMatrix (
    SparseMatrix,
    sparseMatrix,
    set,
    get,
    add,
    transpose,
    fromList
) where

import qualified DataStructures.Dictionary.AVLDictionary as D -- Asumiendo que tienes un módulo de diccionario AVL
import Data.Maybe (fromMaybe)

-- Representación de un índice de matriz
data Index = Idx Int Int deriving (Eq, Ord, Show)

-- Tipo de datos para representar una matriz dispersa
data SparseMatrix = SM Int Int (D.Dictionary Index Int)

-- Construye una matriz nula de dimensiones dadas
sparseMatrix :: Int -> Int -> SparseMatrix
sparseMatrix fil col = SM fil col D.empty

-- Función privada: obtiene el valor en un índice dado
value :: SparseMatrix -> Index -> Int
value (SM _ _ dict) idx = fromMaybe 0 (D.valueOf idx dict)

-- Función privada: actualiza un valor dado un índice
update :: SparseMatrix -> Index -> Int -> SparseMatrix
update (SM fil col dict) idx val
    | val == 0  = SM fil col (D.delete idx dict)
    | otherwise = SM fil col (D.updateOrInsert idx (const val) val dict)

-- Función privada: construye un índice con validación
index :: SparseMatrix -> Int -> Int -> Index
index (SM filMax colMax _) fil col
    | fil < 0 || col < 0 || fil >= filMax || col >= colMax = error "Index out of bounds"
    | otherwise                                            = Idx fil col

-- Función pública: actualiza el valor en una posición (valida el índice)
set :: SparseMatrix -> Int -> Int -> Int -> SparseMatrix
set sm fil col = update sm (index sm fil col)

-- Función pública: obtiene el valor de una posición (valida el índice)
get :: SparseMatrix -> Int -> Int -> Int
get sm fil col = value sm (index sm fil col)

-- Suma de dos matrices dispersas
add :: SparseMatrix -> SparseMatrix -> SparseMatrix
add sm1@(SM fil1 col1 dict1) (SM fil2 col2 dict2)
    | fil1 /= fil2 || col1 /= col2 = error "add on matrices with different dimension"
    | otherwise                    = foldr (\(idx, val) acc -> update acc idx (val + value acc idx)) sm1 (D.keysValues dict2) 

-- Traspuesta de una matriz dispersa
transpose :: SparseMatrix -> SparseMatrix
transpose (SM fil col dict) = foldr (\(Idx f c, val) acc -> update acc (Idx c f) val) (sparseMatrix col fil) (D.keysValues dict)

-- Crea una matriz dispersa a partir de una lista [fila, col, val, fila, col, val, ...]
fromList :: Int -> Int -> [Int] -> SparseMatrix
fromList fil col xs
    | length xs `mod` 3 /= 0 = error "The length of the list is not multiple of 3"
    | otherwise              = fromListRec xs
        where 
            fromListRec []         = sparseMatrix fil col
            fromListRec (x:y:z:xs) = set (fromListRec xs) x y z

-- Complejidad esperada: O(n log k), donde n = longitud de la lista / 3, k = elementos distintos insertados.
